pkg/operation/botanist/operatingsystemconfig.go

```diff
@@ -254,7 +255,12 @@ func (b *Botanist) generateCloudConfigExecutorResourcesForWorker(
 		}
 	}
 
-	executorScript, err := ExecutorScriptFn([]byte(oscDataOriginal.Content), hyperkubeImage, kubernetesVersion.String(), kubeletDataVolume, *oscDataOriginal.Command, oscDataOriginal.Units)
+	workerPoolHash, err := gutils.WorkerPoolHash(&worker, b.Shoot.GetInfo().Spec)
```

Comment:
Calculating the hash here in g/g does not consider the additional data that an extension chooses in order to cause a rolling update, for example https://github.com/gardener/gardener-extension-provider-azure/blob/4c69d42c368739fc05c266ce5b7c800f740c52d4/pkg/controller/worker/machines.go#L410. Hence, this approach fixes the reported CRI issue but does not generically solve the cloud-config evaluation problem.

Proposal: Extensions can add the hash to the node object which is then read by the cloud-config executor (similar to the annotation `checksum/cloud-config-data`). Any alternatives, opinions? (/cc @rfranzke @vpnachev)

Reply:
That's a great catch and an absolutely relevant use-case as well. Looking at the code of the AWS provider, for example, I saw that whether a disk is encrypted or not is one of these additional data points. Currently, this would mean when switching disk encryption we would also see the cloud-config being applied to to-be-removed nodes – not sure which implication that would have given that the configuration wouldn't fit the actually attached volume.

Reply:
Extension contract-wise we cannot generally know how the machines are managed. This `WorkerPoolHash` function is an implementation detail and also the usage of MCM is. There might be extensions which manage their nodes without MCM (I don't think this exists today, but I remember discussions in that area w.r.t. metal nodes). This is all theoretical and proceeding on the proposed path might be a pragmatic fix for the time being, but it is conceptually flawed and might or might not break in the future.

Reply:
A more general approach might be letting the `cloud-config-executor` observe a specific label or annotation on its `Node` resource and stop updating the `cloud-config` if such label/annotation is present. The `Worker` extension controller can become responsible for labelling/annotating `Node`s that it plans to terminate. This can potentially be put into the `genericactuator` so that all provider extensions using it just need revendoring.
I haven't though this through fully, you might want to explore this a bit more.

Reply:
Thanks for the additional suggestions! I think the main problem we're facing here is that any mechanism that is done asynchronously is prone to a race between the downloader already updating the newest cloud-config and the detection mechanism being set on all the nodes (which could be many, like 200 or whatnot). I therefore tried to put it into the cloud-config content, such that we don't run into these issues. Do you see a different way to avoid this?

Reply:
- The function `botanist.DeployManagedResourceForCloudConfigExecutor` is deployed before `botanist.DeployWorker` in the shoot reconciliation flow. It has access to the checksum of the executor data script that is later downloaded on the nodes.
- We could introduce a new field to the `extensionsv1alpha1.Worker` resource and use it to let gardenlet populate this checksum.
- `Worker` extension controllers could optionally read this field and annotate `Node`s with this checksum if they "allow" them being updated. Most notably, they wouldn't annotate `Node`s that are about to be deleted in rolling updates.
- In the executor script, we could now read this annotation from the respective `Node` resources:
  - If the annotation is not present (due to backwards-compatibility, or even purposefully omitted if extension decide not to use/need this mechanism) then we just execute/run the new script just like today.
  - If the value does not match then the new script is NOT executed/ran.
  - If the value matches with the checksum of the downloaded (new) executor script then the new script is executed/ran.
- Note that the `botanist.WaitUntilCloudConfigUpdatedForAllWorkerPools` function would need adaptation.